### PR TITLE
Some proposal after functional testing

### PR DESCRIPTION
**Pull request by adrienpeiffer**
Releted to https://github.com/OCA/queue/pull/1

### CODE REVIEW

File: queue_job/models/queue_job.py [hunk 1]
```diff
@@ -14,9 +14,18 @@
 
 _logger = logging.getLogger(__name__)
 
+from odoo.models import MetaModel
+
 
 def channel_func_name(method):
-    return '<%s>.%s' % (method.im_class._name, method.__name__)
+    method_class = False
+    for key, value in method.func_globals.iteritems():
```
Comment (guewen) [inline queue_job/models/queue_job.py:22]: Couldn't you use `_is_model_method` instead of iterating over the globals?

File: queue_job/job.py [hunk 2]
```diff
@@ -397,20 +405,33 @@ def store(self):
...
-            self.job_model.sudo().create(vals)
-
-    def db_record(self):
-        return self.db_record_from_uuid(self.env, self.uuid)
+            if use_new_cursor:
+                cr = registry(self.job_model.env.cr.dbname).cursor()
+                self.job_model.with_env(
+                    self.job_model.db_record.env(cr=cr)).sudo().create(vals)
+                cr.commit()
+                cr.close()
```
Comment (guewen) [inline queue_job/job.py]: The close should be in a `finally`.
You could factor this out with a context manager

```python

@contextmanager
def use_new_env(self, env):
    cr = registry(env).cursor()
    new_env = env(cr=cr)
    try:
        yield new_env
    except:
        cr.rollback()
        raise
    else:
        cr.commit()
    finally:
        cr.close()

File: queue_job/job.py [hunk 3]
```diff
@@ -397,20 +405,33 @@ def store(self):
...
 
-        db_record = self.db_record()
+        db_record = self.db_record(use_new_cursor=use_new_cursor)
         if db_record:
-            db_record.write(vals)
+            if use_new_cursor:
+                cr = registry(db_record.env.cr.dbname).cursor()
+                db_record.with_env(db_record.env(cr=cr)).write(vals)
+                cr.commit()
+                cr.close()
```
Comment (guewen) [inline queue_job/job.py]: The close should be in a finally.

File: queue_job/job.py [hunk 4]
```diff
@@ -244,9 +246,15 @@ def enqueue(cls, func, args=None, kwargs=None,
...
-    def db_record_from_uuid(env, job_uuid):
+    def db_record_from_uuid(env, job_uuid, use_new_cursor=False):
         model = env['queue.job'].sudo()
+        if use_new_cursor:
+            cr = registry(model.env.cr.dbname).cursor()
+            model = model.with_env(model.env(cr=cr))
         record = model.search([('uuid', '=', job_uuid)], limit=1)
+        if use_new_cursor:
+            cr.commit()
+            cr.close()
```
Comment (guewen) [inline queue_job/job.py]: The close should be in a finally.

File: queue_job/models/queue_job.py (hunk 1 above)
Comment (guewen) [reply queue_job/models/queue_job.py:22]: I saw your commit's comment "im_class seems not accessible". What is the type of `method` in your use case?

**Review by guewen — CHANGES_REQUESTED** [review]
Thanks for the test and changes!

Out of curiosity, what was the need for the new cursor?

I made some comments, mostly:

* the new cr stuff can be factored in a single context manager, besides, it must be closed in a `finally` to be sure it is closed properly in any case
* I'm a bit surprised by the issues you had with the `im_class` so I would be glad to have some details

File: queue_job/models/queue_job.py (hunk 1 above)
Comment (adrienpeiffer) [reply queue_job/models/queue_job.py:22]: As your exemple, this is a simple method defined in a class like 

```
class TestQueueJob(models.Model):

    _name = 'test.queue.job'

    name = fields.Char(required=True)

    @job
    def do_job(self):
        return
```

And with this example and your code, the server fails during the initialization.

**Comment by adrienpeiffer** [comment]
Hi @guewen,

Thanks for comments.
About the cursor, [Here](https://github.com/guewen/queue/pull/2/files#diff-c61ef52f6c5b2184f39e2ae518fe7e77L124) you tried to commit the current transaction just before a raise. However, at this step, we should rollback the transaction by raising an exception. So for me the first step was to remove the commit to rollback the transaction properly. But, after removing this one, the storing of the job was also canceled. So I added the possibility to store a job with a new cursor to allow to store the job without rollback on the current transaction.

File: queue_job/models/queue_job.py (hunk 1 above)
Comment (guewen) [inline queue_job/models/queue_job.py]: You are right, the function is not yet bound to the class at this point. Your check does not seem to work better, in my case, the class is not in the globals of the function. It seems we can only check that at runtime when the delay is called.

File: queue_job/job.py [hunk 5]
```diff
@@ -624,16 +645,12 @@ def delay_from_model(*args, **kwargs):
...
-
-    inner_func = func.__func__
     delay_func = delay_from_model
 
-    inner_func.delayable = True
-    inner_func.delay = delay_func
-    inner_func.retry_pattern = retry_pattern
-    inner_func.default_channel = default_channel
+    func.delayable = True
+    func.delay = delay_func
```
Comment (guewen) [inline queue_job/job.py:631]: could you replace `delay_func` by `delay_from_model` and remove the initialization of `delay_func` above?

File: queue_job/models/queue_job.py (hunk 1 above)
Comment (guewen) [reply queue_job/models/queue_job.py:22]: Indeed. Looping over the `func_globals` bothers me but I don't see another way.

**Comment by adrienpeiffer** [comment]
Hi @guewen,

In my last commit, I improved the way to store a job after exception according OCA guidelines.

**Review by guewen — APPROVED** [review]

**Comment by guewen** [comment]
Thanks a lot!

**Comment by guewen** [comment]
Tests are failing

**Pull request merged**